hybrasyl/Throttle.cs

```diff
@@ -0,0 +1,283 @@
+                info = client.ThrottleState[packet.Opcode];
+            }
+
+            Monitor.Enter(info);
+
+            try
+            {
+                Int64 rightnow = DateTime.Now.Ticks;
+                Int64 transmitInterval = (rightnow - info.LastReceived) / TimeSpan.TicksPerMillisecond;
+                Int64 acceptedInterval = (rightnow - info.LastAccepted) / TimeSpan.TicksPerMillisecond;
+                info.PreviousReceived = info.LastReceived;
+                info.LastReceived = rightnow;
+                //Logger.Error($"Interval is {interval} - maximum is {Interval}");
+                info.TotalReceived++;
+
+                if (info.Throttled)
+                {
+                    result = ThrottleResult.Throttled;
+                    if (acceptedInterval > ThrottleDuration && acceptedInterval >= Interval)
+                    {
+                        Logger.Warn($"Transmit interval {transmitInterval}, last accepted {acceptedInterval} - maximum is {Interval}, not throttled");
+                        info.Throttled = false;
+                        info.TotalThrottled = 0;
+                        result = ThrottleResult.ThrottleEnd;
+                        info.PreviousAccepted = info.LastAccepted;
+                        info.LastAccepted = rightnow;
+                        OnThrottleStop(client);
+                    }
+                    else
+                    {
+                        info.TotalThrottled++;
+                        Logger.Warn($"Throttled, count is {info.TotalThrottled}");
+
+                        result = ThrottleResult.Throttled;
+                        if (ThrottleDisconnectThreshold > 0 && info.TotalThrottled > ThrottleDisconnectThreshold)
+                        {
+                            result = ThrottleResult.Disconnect;
+                            OnDisconnectThreshold(client);
+                        }
+                    }
+                }
+                else
+                {
+                    if (acceptedInterval <= Interval)
+                    {
+                        Logger.Warn($"Transmit interval {transmitInterval}, last accepted {acceptedInterval} - maximum is {Interval}, throttled");
+                        info.Throttled = true;
+                        OnThrottleStart(client);
+                        result = ThrottleResult.Throttled;
+                    }
+                    else
+                    {
+                        info.LastAccepted = rightnow;
+                        info.TotalAccepted++;
+                        result = ThrottleResult.OK;
+                    }
+                }
+            }
+            finally
+            {
+                Monitor.Exit(info);
+
+            }
+            return result;
+        }
+    }
+
+    // TODO: add other types of throttles
+
+    public class ThrottleInfo
+    {
+        public Int64 PreviousReceived;
```

Comment:
is int64 necessary?

Reply:
I believe so, since `DateTime.Now.Ticks` is `long`. The other ones, I didn't want to risk ever having the capacity for an overflow.

Reply:
Why not just use these as a DateTime? That makes it much easier to read, and allows for never having to worry about the overflow.

Reply:
DateTime.Now is not particularly useful for millisecond-level accuracy, which is where most throttling would occur.

Reply:
I disagree.

https://msdn.microsoft.com/en-us/library/system.datetime.addmilliseconds(v=vs.110).aspx